src/main/scala/io/opentargets/etl/backend/Expression.scala

```diff
@@ -1,33 +1,288 @@
+                                     binnedDF: DataFrame,
+                                     zscoreDF: DataFrame): DataFrame = {
+
+    val rnaTransposed = transposeDataframe(rnaDF, Seq("ID")).withColumn("unit", lit("TPM"))
+    val binnedTransposed = transposeDataframe(binnedDF, Seq("ID")).withColumn("unit", lit(""))
+    val zscoreTransposed = transposeDataframe(zscoreDF, Seq("ID")).withColumn("unit", lit(""))
+
+    val rna = standardiseBaseline(rnaTransposed).withColumnRenamed("val", "rna")
+    val binned = standardiseBaseline(binnedTransposed).withColumnRenamed("val", "binned")
+    val zscore = standardiseBaseline(zscoreTransposed).withColumnRenamed("val", "zscore")
+
+    val baseExpressions =
+      unionDataframeDifferentSchema(unionDataframeDifferentSchema(rna, binned), zscore)
+
+    val baseExpressionGrouped = baseExpressions
+      .groupBy("Gene", "Tissue")
+      .agg(max(col("rna")).as("rna_val"),
+           max(col("binned")).as("binned_val"),
+           max(col("zscore")).as("zscore_val"),
+           first("unit", ignoreNulls = true).as("unit_val"))
+
+    baseExpressionGrouped
+  }
+
+  /*
+    Given tissue and efo info it generate the correlated info.
+   */
+  private def efoTissueMapping(mapEfos: DataFrame, expressions: DataFrame): DataFrame = {
+    val expressionsRenamed = expressions
+      .withColumnRenamed("_c0", "expressionId")
+      .withColumnRenamed("_c1", "name")
+
+    val mapEfosRenamed = mapEfos.withColumnRenamed("tissue_id", "tissue_internal_id")
+
+    val mappedInfo = mapEfosRenamed
+      .join(expressionsRenamed, col("name") === col("tissue_internal_id"), "full")
+      .withColumn("efoId", when(col("efo_code").isNull, col("name")).otherwise(col("efo_code")))
+      .withColumn("labelNew", when(col("label").isNull, col("name")).otherwise(col("label")))
+
+    mappedInfo
+  }
+
+  /*
+     It generates the list of valid gene and relative labels.
+   */
+  private def selectTissues(tissues: DataFrame, efoTissueMap: DataFrame) = {
+    val normalTissueLabel =
+      tissues.join(efoTissueMap, col("labelNew") === col("Tissue"), "left")
+    val normalTissueExpression =
+      tissues.join(efoTissueMap, col("expressionId") === col("Tissue"), "left")
+
+    val normalTissueWithLabel = normalTissueLabel.unionByName(normalTissueExpression)
+
+    val emptyLabels = normalTissueWithLabel
+      .filter(col("labelNew").isNull)
+      .withColumn("TissueDef", col("Tissue"))
+      .select("Gene", "TissueDef")
+
+    val hasLabels = normalTissueWithLabel
+      .filter(col("labelNew").isNotNull)
+
+    val hasLabelFiltered = hasLabels
+      .withColumn("TissueDef", col("Tissue"))
+      .select("Gene", "TissueDef")
+
+    val missingLabel = emptyLabels.except(hasLabelFiltered)
+
+    val selectMissingRecords = normalTissueWithLabel
+      .join(missingLabel, Seq("Gene"), "right")
+      .where(col("TissueDef") === col("Tissue"))
+      .withColumn("labelDef", col("TissueDef"))
+      .drop("TissueDef")
```

Comment:
Here you can rename the column rather than adding a new column and dropping the old one.

Reply:
yup. Another good catch. Fixing it...